Initialize asyncio event loop before using it

alimirjamali opened the pull request:
Python 3.14 (in Fedora 43) throws RunetimeError if event loop is not initialized before using it.

Resolves: QubesOS/qubes-issues#10188

alimirjamali committed:
Initialize asyncio event loop before using it

Python 3.14 (in Fedora 43) throws RunetimeError if event loop is not
initialized before using it.

Resolves: QubesOS/qubes-issues#10188

reneleonhardt commented:
Nice contribution! What's still needed for Python 3.14?

reneleonhardt commented:
Thank you!
So is there something missing to undraft and review this? 🙂

marmarek commented:
Yes, but at this time "main" branch is open only for changes that are necessary for R4.3.

reneleonhardt commented:
I couldn't find this information in the README, only the `master` branch is being mentioned there.
Maybe you want to make the documentation clearer or introduce a v4.4 integration branch for future features.

marmarek commented:
See https://doc.qubes-os.org/en/latest/developer/releases/version-scheme.html#release-schedule
Anyway, this is not a place for such discussion.

marmarek committed:
Fix getting event loop when initializing Qubes() instance

During qubesd startup, the event loop is already created, but it isn't
running yet. Use get_event_loop(), not get_running_loop().

QubesOS/qubes-issues#10188

marmarek commented:
There is _a lot_ of failures here, most include `asyncio.exceptions.InvalidStateError`. I'm not 100% sure it's related to this PR, but it's very likely.

marmarek committed:
tests: cancel all remaining tasks between tests execution

If old loop still has some tasks, bad things happen (mostly timeouts
waiting for no longer existing objects). While theoretically all tasks should be
completed by the time test ends, ensure that by cancelling them all (and
logging a warning if there were any).

marmarek committed:
tests: fix initializing event loop

Do not create new event loop for every test, that makes various FD
readers and events not working anymore. While theoretically it should be
possible to find and re-register them all, it isn't worth it.

And also, register libvirt event loop only after creating asyncio one.

marmarek committed:
tests: do not use asyncio.run() or new_event_loop() inside tests

asyncio.run() unsets event loop at the end, which breaks any further use of
asyncio. But the QubesTestCase.setUp() creates event loop, so using
get_event_loop() is okay there.
Similarly new_event_loop() is harmful.

This reverts significant part of 7e4a837e "Initialize asyncio event loop
before using it".

Fixes: 7e4a837e "Initialize asyncio event loop before using it"

ben-grande commented:
- Please use `Fixes: https://github.com/qubesos/qubes-issues/issues/10820` on the commit message. I have assigned you (Marek) to that issue
- Small lint/black error, but unittests succeeded
- Last deployed OpenQA run seems decent, only standard failures (didn't inspect logs, just skimmed through): https://openqa.qubes-os.org/tests/overview?distri=qubesos&version=4.3&build=2026060215-devel

marmarek commented:
In the end, not much left from the original commit. I'll cleanup the history and open new PR.

marmarek commented:
#821
